modem-manager: add Quectel secure boot status AT command

ivanmikh opened the pull request:
Type of pull request:

- [ ] New plugin (Please include [new plugin checklist](https://github.com/fwupd/fwupd/wiki/New-plugin-checklist))
- [ ] Code fix
- [x] Feature
- [ ] Documentation

Add Quectel's specific AT command to check secure boot status of the modem firmware:
AT+QSECBOOT="status"

Typical responses are:
+QSECBOOT: "status",1
+QSECBOOT: "status",0

Firmware versions comparison method should still be used for EM05 modems since it does not expose AT port to the system.

superm1 commented on plugins/modem-manager/fu-mm-device.c:
```diff
@@ -555,10 +736,7 @@ fu_mm_device_probe_default(FuDevice *device, GError **error)
-	    g_str_has_prefix(fu_device_get_name(device), "QUECTEL")) {
-		fu_mm_device_probe_quectel_flags(self);
-	}
+	fu_mm_device_probe_quectel_flags(self);
```
Why not just look for the name like before as a pre-requisite?

ivanmikh replied:
> Why not just look for the name like before as a pre-requisite?

Name is different for every module unfortunately, EG25G is `QUECTEL Mobile Broadband Module`, EM05-G is `EM05-G`, EM160R-GL is `EM160R GL`, etc.

But I agree that it is good to have a check for Quectel devices, I will add one based on VIDs for both `+GETFWBRANCH` and `+QSECBOOT`. There are only two Quectel VIDs for now: `PCI:1eac` and `USB:2c7c`

DylanVanAssche replied:
Can this also be used to turn off secure boot? like `AT+QSECBOOT="secure_boot, 0"`? If secure boot blocks switching branches, we need to handle that part properly.

ivanmikh replied:
> Can this also be used to turn off secure boot? like AT+QSECBOOT="secure_boot, 0"? If secure boot blocks switching branches, we need to handle that part properly.

No, unfortunately. Flashing secure boot firmware blows an efuse, so there is no way to disable it and modem can only be flashed with signed firmware after that.

ivanmikh commented:
So `AT+GETFWBRANCH` is only being run for modems with `ModemManagerBranchAtCommand` quirk set. Currently its only EG25-G.

Also I moved secure boot status to `fu_mm_device_setup` and added a VID check.

@DylanVanAssche please take a look at the `fu_mm_device_setup_branch_at` changes. I changed the logic a little bit so it does not open AT port if branch_at command is not set.

DylanVanAssche commented:
> So `AT+GETFWBRANCH` is only being run for modems with `ModemManagerBranchAtCommand` quirk set. Currently its only EG25-G.
> 
> Also I moved secure boot status to `fu_mm_device_setup` and added a VID check.
> 
> @DylanVanAssche please take a look at the `fu_mm_device_setup_branch_at` changes. I changed the logic a little bit so it does not open AT port if branch_at command is not set.

Looks fine, if the self->at_branch variable is not set because the quirk is not available, it is not needed to run that method.

hughsie commented on plugins/modem-manager/fu-mm-device.c:
```diff
@@ -1742,45 +1730,112 @@ fu_mm_device_attach(FuDevice *device, FuProgress *progress, GError **error)
+						   (FuDeviceLockerFunc)fu_mm_device_io_close,
+						   &error_local);
+		if (locker == NULL)
+			g_debug("failed to open AT port: %s", error_local->message);
```
same here

hughsie commented on plugins/modem-manager/fu-mm-device.c:
```diff
@@ -1742,45 +1730,112 @@ fu_mm_device_attach(FuDevice *device, FuProgress *progress, GError **error)
+	FuMmDevice *self = FU_MM_DEVICE(device);
+
+	vendors = fu_device_get_vendor_ids(device);
+	if (vendors != NULL && vendors->len != 0) {
```
you don't need this line; vendors has to be non-NULL and the loop does the len check for you

ivanmikh commented on plugins/modem-manager/fu-mm-device.c:
```diff
@@ -1742,45 +1730,112 @@ fu_mm_device_attach(FuDevice *device, FuProgress *progress, GError **error)
-		if (!fu_mm_device_at_cmd(self, self->branch_at, TRUE, &error_branch))
-			g_debug("unable to get firmware branch: %s", error_branch->message);
+	if (locker == NULL)
+		g_debug("failed to open AT port: %s", error_local->message);
```
but it is followed by `else if`, so if locker is NULL it prints the error_local message and then function returns

DylanVanAssche commented on plugins/modem-manager/fu-mm-device.c:
```diff
@@ -1742,45 +1730,112 @@ fu_mm_device_attach(FuDevice *device, FuProgress *progress, GError **error)
-		if (!fu_mm_device_setup_branch_at(self, error))
-			return FALSE;
-	}
+	fu_mm_device_setup_branch_at(self);
```
@ivanmikh Since you said that you cannot disable secboot at all, branches should be blocked IMO since you can't switch at all. Or at least disable the `community` option if Quectel decides to use branches at some point.

DylanVanAssche replied:
Or the community one should be able to sign their firmware?

hughsie replied:
ack; if the device is signed then my definition it can't support branch-switching.

ivanmikh replied:
okay, i added a check for signed firmware before running firmware branch AT command

hughsie commented on plugins/modem-manager/fu-mm-device.c:
```diff
@@ -1755,31 +1756,94 @@ fu_mm_device_setup_branch_at(FuMmDevice *self, GError **error)
+
+		if (!fu_mm_device_at_cmd(self, "AT+QSECBOOT=\"status\"", TRUE, &error_local))
+			g_debug("failed to send AT+QSECBOOT command: %s", error_local->message);
+	} else {
+		const gchar *version = fu_device_get_version(FU_DEVICE(self));
+		g_autofree gchar *name = NULL;
```
I think if AT+QSECBOOT fails, we want to fall back to the table; i.e. return TRUE if fu_mm_device_at_cmd succeeded, otherwise fall back to the lower parts. I've added you to the committers, so if you push your branch to the origin server I can fix it up for you.

ivanmikh replied:
thanks! just pushed the latest edit to the origin

hughsie commented on plugins/modem-manager/fu-mm-device.c:
```diff
@@ -1755,31 +1756,94 @@ fu_mm_device_setup_branch_at(FuMmDevice *self, GError **error)
+	GPtrArray *vendors;
+	FuMmDevice *self = FU_MM_DEVICE(device);
+
+	vendors = fu_device_get_vendor_ids(device);
+	for (guint i = 0; i < vendors->len; i++) {
```
Just use fu_device_has_vendor_id() here please.

Biktorgj commented:
Just my two cents, the EG25-G, from version `EG25GGBR07A07M2G_30.003.30.003` and higher, has a different command to check secure boot:
`AT+QCFG="secbootstat"`
Possible responses seem to be:
`+QCFG: "secbootstat",0`
`+QCFG: "secbootstat",1`

ivanmikh commented:
> Just my two cents, the EG25-G, from version `EG25GGBR07A07M2G_30.003.30.003` and higher, has a different command to check secure boot: `AT+QCFG="secbootstat"` Possible responses seem to be: `+QCFG: "secbootstat",0` `+QCFG: "secbootstat",1`

Is this one available on community firmware?

Biktorgj commented:
No, this is only available in stock, but I can add it to the community firmware if needed

ivanmikh commented:
> No, this is only available in stock, but I can add it to the community firmware if needed (right now with the community firmware it will just return "OK")

Its fine. Thanks for bringing it up, will add it to the commit as well

ivanmikh commented:
Hey,
I have updated the logic, so if AT commands fails it falls back to the table comparison. Also added another AT command for EG25-G

hughsie reviewed (approved):
Thanks!
